### PR TITLE
Firecracker Add jailer support for firecracker

### PR DESCRIPTION
**Pull request by mcastelino**
Firecracker provides a jailer to constrain the VMM. Use this
jailer to launch the firecracker VMM instead of launching it
directly from the kata-runtime.

The jailer will ensure that the firecracker VMM will run
in its own network and mount namespace. All assets required
by the VMM have to be present within these namespaces.
The assets need to be copied or bind mounted into the chroot
location setup by jailer in order for firecracker to access
these resouces. This includes files, device nodes and all
other assets.

Jailer automatically sets up the jail to have access to
kvm and vhost-vsock.

If a jailer is not available (i.e. not setup in the toml)
for a given hypervisor the runtime will act as the jailer.

Also enhance the hypervisor interface and unit tests to
include the network namespace. This allows the hypervisor
to choose how and where to lauch the VMM process, vs
virtcontainers directly launching the VMM process.

Fixes: #1129

Signed-off-by: Manohar Castelino <manohar.r.castelino@intel.com>

### CODE REVIEW

**Comment by mcastelino** [comment]
Cannot be merged unless jailer bug is fixed https://github.com/firecracker-microvm/firecracker/issues/1089

**Comment by mcastelino** [comment]
We also need to fix kata network namespace logic to be able to merge this PR https://github.com/kata-containers/runtime/issues/1664

File: virtcontainers/fc.go [hunk 1]
```diff
@@ -242,17 +289,53 @@ func (fc *firecracker) fcInit(timeout int) error {
...
+		fc.netNSPath = networkNS.NetNsPath
+	}
+
+	var args []string
+	var cmd *exec.Cmd
+
+	if fc.jailed {
+		args = []string{
+			"--id", fc.id,
+			"--node", "0", //FIXME: Comprehend NUMA topology or explicit ignore
```
Comment (bergwolf) [inline virtcontainers/fc.go]: For now, don't set it unless the local allocator (kubelet etc.) tells us which NUMA node to use?

**Comment by mcastelino** [comment]
If firecracker accepts https://github.com/firecracker-microvm/firecracker/pull/1093 we can support jailer in Kata.

@devimc @amshinde I had the change the bind mount setup in Kata from private to slave in order for things to work correctly. So we need to change our bindmount logic to choose the right type of bind mount on a case by case basis vs always choosing private.

See 8e2ce77d6c3fd26e26ae0dae8c84244ba451e673 
I will need to fix this more cleanly in kata or create my own bindmount routine that does right for jailer.

**Comment by devimc** [comment]
@mcastelino 

> @devimc @amshinde I had the change the bind mount setup in Kata from private to slave in order for things to work correctly. So we need to change our bindmount logic to choose the right type of bind mount on a case by case basis vs always choosing private.

no way, see https://github.com/kata-containers/runtime/issues/777 and https://github.com/kata-containers/tests/blob/master/integration/docker/cp_test.go#L74

**Comment by devimc** [comment]
@mcastelino 

> @devimc @amshinde I had the change the bind mount setup in Kata from private to slave in order for things to work correctly. So we need to change our bindmount logic to choose the right type of bind mount on a case by case basis vs always choosing private.

Currently, that's not possible, because of 9p. When overlay is used as storage driver, kata runtime creates a new bind mount point to the merged directory, that way this directory can be shared with the VM through 9p. By default the mount propagation is shared, that means mount events are propagated, but umount events not, to deal with this problem and to avoid left mount points in the host once container finishes, the mount propagation of bind mounts should be set to private.

@mcastelino **if you change the mount propagation to slave right now, you will re-introduce a bug. We can switch slave, once we support virtio-fs and deprecate 9p**

@mcastelino does it make sense?

cc @egernst  @amshinde

**Comment by mcastelino** [comment]
> By default the mount propagation is shared, that means mount events are propagated, but umount events not, to deal with this problem and to avoid left mount points in the host once container finishes, the mount propagation of bind mounts should be set to private.

@devimc when you say unmount are not propagated, do you mean umount from within the pivot_root/chroot? In the case of firecracker all the mounts and umount happen from outside the pivot_root. So the setup and cleanup is done by kata. 

When you say mounts are left being, is this the mounts done by docker as part of docker cp or the mounts created by Kata?

I am trying to understand why I am not seeing the issue when I use SLAVE.

**Comment by devimc** [comment]
@mcastelino 

> @devimc when you say unmount are not propagated, do you mean umount from within the pivot_root/chroot? 

no, when the mount propagation is slave/shared, the mount point created outside the pivot_root are propagated (visible within it), but umount event not, because we are creating a bind mount to the merged (overlay) directory. Once we support virtio-fs and deprecated 9p, won't be needed to create a bind mount point, since the fuse daemon will be able to monitor the merged directory directly.

> When you say mounts are left being, is this the mounts done by docker as part of docker cp or the mounts created by Kata?

yes, for a strange reason that I still don't understand, `docker cp` remounts all the volumes, once it finishes, those mounts are umounted by docker, unfortunately for us, remounts are propagated within the bind mount that we create, but umount events are not propagated, that way mounts are left behind causing bugs.

**Comment by mcastelino** [comment]
@jodh-intel the struct does not pass the maligned linter check and I am clueless at this point, as to how to fix this.

File: virtcontainers/fc.go (hunk 1 above)
Comment (bergwolf) [reply virtcontainers/fc.go]: I see. Let's have it for now then.

**Comment by raravena80** [comment]
@mcastelino any updates? Thx!

**Comment by mcastelino** [comment]
> @mcastelino any updates? Thx!

@raravena80 this is actually ready for merge. Not sure what is happening here with the CI. 

The change needed for firecracker has been merged upstream https://github.com/firecracker-microvm/firecracker/pull/1093

However I do not think there is a release with this bugfix. Maybe that is why the CI is failing.

So I think we will need to first update firecracker to use a release with this bugfix and then merge this PR.

File: virtcontainers/fc.go [hunk 2]
```diff
@@ -143,9 +157,47 @@ func (fc *firecracker) trace(name string) (opentracing.Span, context.Context) {
 	return span, ctx
 }
 
+// bindMount bind mounts a source in to a destination. This will
+// do some bookkeeping:
+// * evaluate all symlinks
+// * ensure the source exists
+// * recursively create the destination
```
Comment (amshinde) [inline virtcontainers/fc.go]: Can you add a note to specify that you are adding this to override the bind-mount behaviour in Kata to use a slave mount instead?

Comment (mcastelino) [reply virtcontainers/fc.go]: @amshinde you are right. This is only for firecracker specific. Also it does not impact firecracker as it does not use 9p today.

**Comment by amshinde** [comment]
Blocked on firecracker next release.

**Comment by caoruidong** [comment]
ping @mcastelino

**Comment by mcastelino** [comment]
> ping @mcastelino
@caoruidong @amshinde 

This was a release Firecracker v0.17.0 4 days ago which has the required code changes. Let me retest with that release and then slag this PR for merge.

**Comment by mcastelino** [comment]
This can be merged once https://github.com/kata-containers/runtime/pull/1813 has passed CI and merged.

**Comment by mcastelino** [comment]
This now passes local testing. So this can be merged once https://github.com/kata-containers/runtime/pull/1813 is merged.

With this you can support firecracker 0.17.0 with and without jailer.
The current default will be without jailer.

This will be followed up by a PR to to enable jailer by default in the toml file.

**Comment by mcastelino** [comment]
Needs https://github.com/kata-containers/tests/pull/1757 to merge first as it is built on top of that PR.

https://github.com/kata-containers/tests/pull/1757 will ensure the 0_17_0 change passes CI.
Then this one will add jailer support **and enable it by default**

/cc @egernst @chavafg

File: virtcontainers/fc.go [hunk 3]
```diff
@@ -314,11 +436,27 @@ func (fc *firecracker) client() *client.Firecracker {
...
-		fc.fcClient = fc.newFireClient()
+	if fc.connection == nil {
+		fc.connection = fc.newFireClient()
 	}
 
-	return fc.fcClient
+	return fc.connection
+}
+
+func (fc *firecracker) fcJailResource(src, dst string) (string, error) {
```
Comment (jodh-intel) [inline virtcontainers/fc.go]: Is it worth adding checks for these params to ensure they are not `""`?

File: virtcontainers/fc.go [hunk 4]
```diff
@@ -327,37 +465,50 @@ func (fc *firecracker) fcSetBootSource(path, params string) error {
...
 	src := &models.BootSource{
-		KernelImagePath: &path,
+		KernelImagePath: &kernelPath,
 		BootArgs:        params,
 	}
 	bootSrcParams.SetBody(src)
 
-	_, err := fc.client().Operations.PutGuestBootSource(bootSrcParams)
-	return err
+	if _, err := fc.client().Operations.PutGuestBootSource(bootSrcParams); err != nil {
```
Comment (jodh-intel) [inline virtcontainers/fc.go]: I'm not clear what's wrong with the old code - it's shorter and clearer to me...?

File: virtcontainers/fc.go [hunk 5]
```diff
@@ -510,6 +666,36 @@ func (fc *firecracker) createDiskPool() error {
 	return nil
 }
 
+func (fc *firecracker) umountResource(jailedPath string) {
+	hostPath := filepath.Join(fc.jailerRoot, jailedPath)
+	syscall.Unmount(hostPath, syscall.MNT_DETACH)
```
Comment (jodh-intel) [inline virtcontainers/fc.go]: The lazy unmount could still fail if the path is incorrect somehow, so I'd log the return value or return it anyway.

Comment (mcastelino) [reply virtcontainers/fc.go]: done

File: virtcontainers/fc.go (hunk 4 above)
Comment (mcastelino) [reply virtcontainers/fc.go]: done

File: virtcontainers/fc.go (hunk 3 above)
Comment (mcastelino) [reply virtcontainers/fc.go]: done

**Comment by mcastelino** [comment]
@jodh-intel all review comments have been addressed and code rebased and tested with latest kata and firecracker 0.17.0

/cc @jcvenegas @devimc @egernst

File: arch/amd64-options.mk [hunk 6]
```diff
@@ -13,6 +13,8 @@ QEMUCMD := qemu-system-x86_64
 
 # Firecracker binary name
 FCCMD := firecracker
+# Firecracker's jailer binary name
+FCJAILERCMD := jailer
```
[the remark this answers is not in the document]
Comment (mcastelino) [reply arch/amd64-options.mk]: I was thinking about that. But as of now we do not really have a real jailer for any other Hypervisor.  The fc jailer if generalized may be able to support other hypervisors. But looking at the feature set I think we can even use `runc` as a jailer.

File: virtcontainers/fc.go [hunk 7]
```diff
@@ -151,10 +203,32 @@ func (fc *firecracker) createSandbox(ctx context.Context, id string, hypervisorC
...
+	// a specific location and that location needs to have
+	// exec permission (i.e. should not be mounted noexec, e.g. /run, /var/run)
+	// Also unix domain socket names have a hard limit
+	// #define UNIX_PATH_MAX   108
+	// Keep it short and live within the jailer expected paths
+	// <chroot_base>/<exec_file_name>/<id>/
+	// Also jailer based on the id implicitly sets up cgroups under
+	// <cgroups_base>/<exec_file_name>/<id>/
+	hypervisorName := filepath.Base(hypervisorConfig.HypervisorPath)
+	fc.chrootBaseDir = "/var/lib/vc" //store.ConfigStoragePath //This mount has exec perms
```
[the remark this answers is not in the document]
Comment (mcastelino) [reply virtcontainers/fc.go]: Will fix this.

File: virtcontainers/fc.go [hunk 8]
```diff
@@ -151,10 +203,32 @@ func (fc *firecracker) createSandbox(ctx context.Context, id string, hypervisorC
...
+
+	fc.vmPath = filepath.Join(fc.chrootBaseDir, hypervisorName, fc.id)
+	fc.jailerRoot = filepath.Join(fc.vmPath, "root") // auto created by jailer
+	fc.socketPath = filepath.Join(fc.jailerRoot, fcSocket)
+
+	// So we need to repopulate this at startSandbox where it is valid
+	fc.netNSPath = networkNS.NetNsPath
+
+	// Till we create lower privileged kata user run as root
```
[the remark this answers is not in the document]
Comment (mcastelino) [reply virtcontainers/fc.go]: Will add an issue.

File: virtcontainers/fc.go [hunk 9]
```diff
@@ -510,6 +671,39 @@ func (fc *firecracker) createDiskPool() error {
 	return nil
 }
 
+func (fc *firecracker) umountResource(jailedPath string) {
+	hostPath := filepath.Join(fc.jailerRoot, jailedPath)
+	err := syscall.Unmount(hostPath, syscall.MNT_DETACH)
+	if err != nil {
+		fc.Logger().WithField("umountResource failed", err).Info()
```
[the remark this answers is not in the document]
Comment (mcastelino) [reply virtcontainers/fc.go]: This currently if info as I know some of these will fail on second go around. There is a comment in the caller but not here. Let me add another comment here.

File: virtcontainers/fc.go [hunk 10]
```diff
@@ -510,6 +671,39 @@ func (fc *firecracker) createDiskPool() error {
...
+
+	//Run through the list second time as may have bindmounted
+	//to the same location twice. In the future this needs to
+	//be tracked so that we do not do this blindly
+	for i := 0; i < fcDiskPoolSize; i++ {
+		fc.umountResource(fcDriveIndexToID(i))
+	}
+
+	fc.Logger().WithField("cleaningJail", fc.vmPath).Error()
+	if err := os.RemoveAll(fc.vmPath); err != nil {
```
[the remark this answers is not in the document]
Comment (mcastelino) [reply virtcontainers/fc.go]: Can you expand a bit on this comment. Not sure I understood.

File: virtcontainers/fc.go [hunk 11]
```diff
@@ -510,6 +671,39 @@ func (fc *firecracker) createDiskPool() error {
 	return nil
 }
 
+func (fc *firecracker) umountResource(jailedPath string) {
```
[the remark this answers is not in the document]
Comment (mcastelino) [reply virtcontainers/fc.go]: Can do. The error can be ignored by caller.

File: virtcontainers/fc.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (mcastelino) [reply virtcontainers/fc.go]: https://github.com/kata-containers/runtime/issues/1869

File: virtcontainers/fc.go [hunk 12]
```diff
@@ -239,17 +313,65 @@ func (fc *firecracker) fcInit(timeout int) error {
...
-	args := []string{"--api-sock", fc.socketPath}
+	if fc.config.JailerPath != "" {
+		fc.jailed = true
+	}
+
+	// Fetch sandbox network to be able to access it from the sandbox structure.
+	var networkNS NetworkNamespace
+	if err := fc.store.Load(store.Network, &networkNS); err == nil {
+		if networkNS.NetNsPath == "" {
+			fc.Logger().WithField("NETWORK NAMESPACE NULL", networkNS).Error()
```
[the remark this answers is not in the document]
Comment (mcastelino) [reply virtcontainers/fc.go]: fixed

**Comment by jcvenegas** [comment]
rhel:
Job fail seems not related @grahamwhaley @GabyCT heads up... restarting

```
bats integration/entropy/entropy_time.bats
1..1
not ok 1 measured time for /dev/random
# (in test file integration/entropy/entropy_time.bats, line 29)
#   `[ "$measured_time" -le "$expected_time" ]' failed
# 
# 
make: *** [entropy] Error 1
Failed at 40: sudo -E PATH="$PATH" bash -c "make test"
```

**Comment by mcastelino** [comment]
Also the centos job is failing due to vsock kernel module being absent

```
Stderr: docker: Error response from daemon: OCI runtime create failed: [PUT /actions][400] createSyncActionBadRequest  &{FaultMessage:Cannot create vsock device. VhostOpen(VhostOpen(Os { code: 19, kind: Other, message: No such device }))}: unknown.
```

**Comment by grahamwhaley** [comment]
> rhel:
> Job fail seems not related @grahamwhaley @GabyCT heads up... restarting
> 
> ```
> bats integration/entropy/entropy_time.bats
> 1..1
> not ok 1 measured time for /dev/random
> # (in test file integration/entropy/entropy_time.bats, line 29)
> #   `[ "$measured_time" -le "$expected_time" ]' failed
> # 
> # 
> make: *** [entropy] Error 1
> Failed at 40: sudo -E PATH="$PATH" bash -c "make test"
> ```

thanks @jcvenegas and, :irony: thanks :/irony: bats, for not giving us any useful actual details of the values of the comparison that failed :-(
iirc, I asked/noted at the time that it would be really helpful if that entropy time test could print out the values upon failure to aid debug - but - I'm not sure if that is something easily done with BATS? - can you look into that @GabyCT ?

**Comment by mcastelino** [comment]
@GabyCT the FC CI is passing on ubuntu for me by following the steps as below. 

```
export PATH=${GOPATH}/bin:${PATH}
DEBUG=true .ci/setup.sh
export KATA_DEV_MODE="false"
export KATA_HYPERVISOR="firecracker"
export CI="true"
export CI_JOB="FIRECRACKER"
export RUNTIME="kata-runtime"
make docker
```

**Comment by mcastelino** [comment]
The tests need update to allow the jailed version of firecracker to be detected

https://github.com/kata-containers/tests/pull/1803

**Comment by GabyCT** [comment]
@mcastelino so it seems that there is an issue with `initrd`

**Comment by GabyCT** [comment]
@mcastelino , the issue with `initrd` seems not related with this PR as I am seeing this failure in other jobs.

**Comment by raravena80** [comment]
🎉 🎈

**Pull request merged**